### PR TITLE
Fix sample test failure because of the type information in the pipelineparam

### PR DESCRIPTION
**Pull request by gaoning777**
op_to_template resolve the raw arguments by mapping to the argument_inputs but the argument_inputs lost the type information.

Add back the type information when argument_input pipeline_param is created.
Also, adjust the TypeMeta.from_dict_or_str to convert a json string back to TypeMeta

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/972)

### CODE REVIEW

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -223,9 +223,12 @@ def _process_args(self, raw_args, argument_inputs):
       matches += _match_serialized_pipelineparam(str(processed_args[i]))
       unsanitized_argument_inputs = {}
       for x in list(set(matches)):
-        sanitized_str = str(dsl.PipelineParam(K8sHelper.sanitize_k8s_name(x[1]), K8sHelper.sanitize_k8s_name(x[0]), x[2]))
-        unsanitized_argument_inputs[sanitized_str] = str(dsl.PipelineParam(x[1], x[0], x[2]))
-
+        if len(x) == 3 or (len(x) == 4 and x[3] == ''):
```
Comment (hongye-sun) [inline sdk/python/kfp/compiler/compiler.py]: nit: could you add some comments on what's the expected data and length on the x? It's hard to understand what the code does from hardcoded indexes.

File: sdk/python/kfp/dsl/_metadata.py [hunk 2]
```diff
@@ -48,10 +48,15 @@ def to_dict_or_str(self):
   @staticmethod
   def from_dict_or_str(json):
```
Comment (hongye-sun) [inline sdk/python/kfp/dsl/_metadata.py]: Better to add some comments on what kinds of data can be passed to the method? It seems that the supported data is not just a json.

File: sdk/python/kfp/dsl/_metadata.py [hunk 3]
```diff
@@ -48,10 +48,15 @@ def to_dict_or_str(self):
...
     type_meta = TypeMeta()
+    if isinstance(json, str) and '{' in json:
```
Comment (hongye-sun) [inline sdk/python/kfp/dsl/_metadata.py]: It's kind of hacky. Could you explain why this is needed?

File: sdk/python/kfp/dsl/_metadata.py [hunk 4]
```diff
@@ -48,10 +48,15 @@ def to_dict_or_str(self):
...
+      import ast
+      json = ast.literal_eval(json)
     if isinstance(json, dict):
       if not _check_valid_type_dict(json):
         raise ValueError(json + ' is not a valid type string')
       type_meta.name, type_meta.properties = list(json.items())[0]
+      # Convert possible OrderedDict to dict
+      type_meta.properties = dict(type_meta.properties)
```
Comment (hongye-sun) [inline sdk/python/kfp/dsl/_metadata.py]: This code will always override the properties that has been set on the line above. Is it correct?

File: sdk/python/tests/dsl/pipeline_param_tests.py [hunk 5]
```diff
@@ -49,4 +49,6 @@ def test_extract_pipelineparam(self):
     self.assertListEqual([p1, p2, p3], params)
     payload = [str(p1) + stuff_chars + str(p2), str(p2) + stuff_chars + str(p3)]
     params = _extract_pipelineparams(payload)
-    self.assertListEqual([p1, p2, p3], params)
\ No newline at end of file
+    self.assertListEqual([p1, p2, p3], params)
+
+  #TODO: add more unit tests to cover real type instances
```
Comment (hongye-sun) [inline sdk/python/tests/dsl/pipeline_param_tests.py]: nit: as discussed offline, it's better to add a basic e2e test in basic sample folder for the type check feature, which will be verified in the e2e presubmit. It's fine to add it later.

File: sdk/python/kfp/dsl/_pipeline_param.py [hunk 6]
```diff
@@ -48,7 +48,13 @@ def _extract_pipelineparams(payloads: str or list[str]):
   matches = []
   for payload in payloads:
     matches += _match_serialized_pipelineparam(payload)
-  return [PipelineParam(x[1], x[0], x[2]) for x in list(set(matches))]
+  pipeline_params = []
+  for x in list(set(matches)):
```
Comment (hongye-sun) [inline sdk/python/kfp/dsl/_pipeline_param.py]: nit: I really think it's not a good coding practice to return a matches from _match_serialized_pipelineparam. It should return a custom type which converts match to strong typed data. It's very hard to maintain the magic number index in the future. You don't need to fix this comment in the PR.

File: sdk/python/kfp/dsl/_metadata.py (hunk 4 above)
Comment (gaoning777) [reply sdk/python/kfp/dsl/_metadata.py]: The reason to convert the OrderedDict to dict is that component module generates the ordereddict and the serialization of the orderedDict is hard to de-serialize when the pipelineparam string is matched to our regex pattern.

File: sdk/python/tests/dsl/pipeline_param_tests.py (hunk 5 above)
Comment (gaoning777) [reply sdk/python/tests/dsl/pipeline_param_tests.py]: Got it. I created an issue to track this: https://github.com/kubeflow/pipelines/issues/974
And, another PR to add some unit tests: https://github.com/kubeflow/pipelines/pull/975

**Comment by hongye-sun** [comment]
/lgtm

None of the comments has to be fixed in the PR. Please feel free to merge the PR and fix the comments in future PRs.

File: sdk/python/kfp/compiler/compiler.py (hunk 1 above)
Comment (gaoning777) [reply sdk/python/kfp/compiler/compiler.py]: Done.

File: sdk/python/kfp/dsl/_metadata.py (hunk 2 above)
Comment (gaoning777) [reply sdk/python/kfp/dsl/_metadata.py]: Done.

File: sdk/python/kfp/dsl/_metadata.py (hunk 3 above)
Comment (gaoning777) [reply sdk/python/kfp/dsl/_metadata.py]: This is to detect whether it is a str or a serialized str of a dict. Does that make sense?

**Comment by gaoning777** [comment]
Will address the PR comments in https://github.com/kubeflow/pipelines/pull/975

File: sdk/python/kfp/dsl/_metadata.py (hunk 3 above)
Comment (gaoning777) [reply sdk/python/kfp/dsl/_metadata.py]: Let me separate the logic into a function called de-serialize

Comment (gaoning777) [reply sdk/python/kfp/dsl/_metadata.py]: Updated to remove the hacking way to detect the serialized string in the new PR.

File: sdk/python/kfp/dsl/_pipeline_param.py (hunk 6 above)
Comment (gaoning777) [reply sdk/python/kfp/dsl/_pipeline_param.py]: make sense. Done in the #975

**Pull request merged**